File: herminedb-driver/src/main/java/org/hermine/internal/db/driver/ConnectionImpl.java
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol,
 * and explicitly declared author of this file if provided.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.hermine.internal.db.driver;

import jdk.incubator.sql2.*;
import reactor.core.scheduler.Scheduler;

import java.util.Map;
import java.util.function.Consumer;

class ConnectionImpl extends SequentialOperationGroup<Object, Object> implements Connection {

    // FIELDS
    private final DataSource dataSource;
    private final Map<ConnectionProperty, Object> properties;

    private final Scheduler scheduler;

    ConnectionImpl(DataSourceImpl ds, Map<ConnectionProperty, Object> properties) {
        super(null, null); // hack as _this_ not allowed. See SimpleOperation constructor
        dataSource = ds;
        this.properties = properties;
        ConnectionProperty execProp = HermineConnectionProperty.SCHEDULER;
        scheduler = (Scheduler) properties.getOrDefault(execProp, execProp.defaultValue());
    }

    @Override
    public Operation<Void> connectOperation() {
        if (!isHeld()) {
            throw new IllegalStateException("TODO");
        }
        return SimpleOperation.<Void>newOperation(this, this, this::hermineConnect);
    }

    @Override
    public Operation<Void> validationOperation(Validation depth) {
        return null;
    }

    @Override
    public Operation<Void> closeOperation() {
        return UnskippableOperation.<Void>newOperation(this, this, this::hermineClose);
    }

    @Override
    public <S, T> AbstractOperationGroup<S, T> operationGroup() {
        return null;
    }

    @Override
    public Transaction transaction() {
        return null;
    }

    @Override
    public ConnectionImpl registerLifecycleListener(ConnectionLifecycleListener listener) {
        return this;
    }

    @Override
    public ConnectionImpl deregisterLifecycleListener(ConnectionLifecycleListener listener) {
        return this;
    }

    @Override
    public Lifecycle getConnectionLifecycle() {
        return null;
    }

    @Override
    public ConnectionImpl abort() {
        return this;
    }

    @Override
    public Map<ConnectionProperty, Object> getProperties() {
        return null;
    }

    @Override
    public ShardingKey.Builder shardingKeyBuilder() {
        return null;
    }

    @Override
    public ConnectionImpl requestHook(Consumer<Long> request) {
        return null;
    }

    @Override
    public ConnectionImpl activate() {
        return this;
    }

    @Override
    public ConnectionImpl deactivate() {
        return this;
    }

    // INTERNAL

    @Override
    protected Scheduler getScheduler() {
        return scheduler;
    }

    private Void hermineConnect(AbstractOperation<Void> op) {
        return null;
    }

    private Void hermineClose(AbstractOperation<Void> op) {
        return null;
    }
}

File: herminedb-driver/src/main/kotlin/org/hermine/internal/db/driver/ConnectionImpl.kt
/*
 * Copyright 2018 HermineDB's author : Frédéric Montariol
 * and explicitly declared author of each file if specified.
 * Use of this source code is governed by the Apache 2.0 license.
 */

package org.hermine.internal.db.driver

import jdk.incubator.sql2.*
import jdk.incubator.sql2.Connection.Lifecycle
import java.time.Duration
import java.util.function.Consumer
import kotlin.coroutines.experimental.CoroutineContext

internal class ConnectionImpl(
        dataSource: DataSourceImpl,
        properties: Map<ConnectionProperty, Any>
) : OperationGroupImpl<Any?, Any>(), Connection {

    override val context: CoroutineContext

    init {
        val execProp = CONTEXT
        context = properties.getOrDefault(execProp, execProp.defaultValue()) as CoroutineContext
    }

    // PUBLIC
    override fun connectOperation(): AbstractOperation<Void?> {
        if (!isHeld()) throw IllegalStateException("TODO")
        return SimpleOperation(this, this, this::hermineConnect)
    }

    override fun validationOperation(depth: Connection.Validation?): Operation<Void> {
        if (!isHeld()) throw IllegalStateException("TODO")
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun closeOperation(): Operation<Void?> {
        if (!isHeld()) throw IllegalStateException("TODO")
        return UnskippableOperation(this, this, this::hermineClose)
    }

    override fun <S : Any?, T : Any?> operationGroup(): OperationGroup<S, T> {
        if (!isHeld()) throw IllegalStateException("TODO")
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun transaction(): Transaction {
        if (!isHeld()) throw IllegalStateException("TODO")
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun registerLifecycleListener(listener: Connection.ConnectionLifecycleListener?): Connection {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun deregisterLifecycleListener(listener: Connection.ConnectionLifecycleListener?): Connection {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getConnectionLifecycle(): Lifecycle {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun abort(): Connection {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getProperties(): MutableMap<ConnectionProperty, Any> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun shardingKeyBuilder(): ShardingKey.Builder {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun requestHook(request: Consumer<Long>?): Connection {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun activate(): Connection {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun deactivate(): Connection {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    // Override Operation methods

    override fun timeout(minTime: Duration): ConnectionImpl {
        super.timeout(minTime)
        return this
    }

    override fun onError(handler: Consumer<Throwable>): ConnectionImpl {
        super.onError(handler)
        return this
    }

    // INTERNAL
    protected fun setLifecycle(next: Lifecycle): ConnectionImpl {
        return this
    }

    private fun hermineConnect(op: AbstractOperation<Void?>): Void? {
        return null
    }

    private fun hermineClose(op: AbstractOperation<Void?>): Void? {
        return null
    }
}
